### ISSUE TITLE
64x64 input r50 model, Caffe conversion gives different result

### ISSUE THREAD

**Issue by WIll-Xu35** [deepinsight/insightface#937]
Hi all,

I'm using tools from this repo [https://github.com/GarrickLin/MXNet2Caffe](url) to convert resnet50 insightface model to caffe.

I've tested the pretrained model, the conversion is success and the result is exactly the same.

However, when I tried it out on my trained model, the inference output is completely different. The command used to train my model is `python -u train.py --network r50 --loss arcface --dataset emore`.

Compared with the original setting, I had two modifications. First, I changed the value of config.ckpt_embedding to False in config.py, in order to save the last fc layer for future finetune. And I pruned fc7 layer afterward and the model's performance is as expected. The second difference is that I changed the dataset.emore.image_shape in config file to (64,64,3) to make the model accept 64x64 face images. And after the execution of json2prototxt.py, I manually changed the input shape to 1, 3, 64, 64. After generate the caffemodel, I can do inference, but the output is not the same as mxnet.

I suspect that changing the input size may lead to the conversion failure. But I have no idea how to solve this problem. Any idea how to make it work for 64x64 input image? @nttstar @GarrickLin

Many thanks.

**Comment by maxuehao** [comment]
You can use mmdnn

**Comment by WIll-Xu35** [comment]
@maxuehao Thank you for your reply. I tried mmdnn several months ago, but I met some dependency problem during installation. I'll give it a try again.

**Comment by WIll-Xu35** [comment]
@maxuehao I've tried mmdnn's docker, update the mmdnn within the docker to the latest version, and then do the conversion. But after the conversion, the output is still not the same. However, I tried to convert the provided pretrained r50 model and the result is almost the same. So it seems that changing he input shape makes everything not working?

**Comment by youthM** [comment]
hi, would you mind releasing your caffe conversion of the provided pretrained r50 model ? @WIll-Xu35

**Comment by WIll-Xu35** [comment]
@youthM 

The model can be downloaded from this link: [https://drive.google.com/open?id=1D6gG2WRvOvUI7NyJLInGl2zBD6zDPcoI](url)

And the problematic caffemodel converted from my own 64x64 model is available at: https://drive.google.com/open?id=1Bc-Lc-LcN1zwof4zuVIZW5Nm5FYQZ44F

If you had extra time and experience with model conversion, could you please take a look at the 64x64 caffemodel, whose output is different from the mxnet model. Many thanks.

**Comment by WIll-Xu35** [comment]
@maxuehao I've tried mmdnn within docker this time. The conversion can be completed, but the output is still not even close to the mxnet model's output. The models are all available in the provided google drive links. If you have time, would you please try to convert the 64x64 model and see if your conversion is successful. Many thanks.

**Comment by youthM** [comment]
> @youthM
> 
> The model can be downloaded from this link: [https://drive.google.com/open?id=1D6gG2WRvOvUI7NyJLInGl2zBD6zDPcoI](url)
> 
> And the problematic caffemodel converted from my own 64x64 model is available at: https://drive.google.com/open?id=1Bc-Lc-LcN1zwof4zuVIZW5Nm5FYQZ44F
> 
> If you had extra time and experience with model conversion, could you please take a look at the 64x64 caffemodel, whose output is different from the mxnet model. Many thanks.

thanks, I'm sorry that I didn't have experience with model conversion.

**Comment by WIll-Xu35** [comment]
I've checked the converted weights, they all seems to be the same.

When checking the output of the hidden layers given the same input, the output of the first 4-5 layers are almost the same. But after that, the outputs starts to diverge and become completely different.

**Comment by SE2AI** [comment]
> Hi all,
> 
> I'm using tools from this repo [https://github.com/GarrickLin/MXNet2Caffe](url) to convert resnet50 insightface model to caffe.
> 
> I've tested the pretrained model, the conversion is success and the result is exactly the same.

hi @WIll-Xu35, I just download your transfered caffemodel from your google drive link, but the inferenced features of mxnet model and caffe model are totally different. 
For the mxnet result, i think its should be ok, it could distinguish different faces. But for the caffe model, it doesn't work. Though the image pre-process and post-process are both same.
Maybe something wrong with my caffemodel's forward script, would you mind share your inference code? lots of thanks!
